### ISSUE TITLE
DateTime.diff and Interval.length can return wrong month counts

### ISSUE THREAD

**Issue by captainmuppet** [moment/luxon#537]
Durations between DateTimes are calculated correctly inside the month or for integer multiples, but the values are wrong if the interval spans multiple months and at least one of them is not a full month.

```js
const { DateTime, Settings } = require('luxon');

Settings.defaultZoneName = 'Europe/Berlin';

const start = DateTime.fromISO('2019-04-06T00:00:00.000+02:00');
const end = DateTime.fromISO('2019-06-01T00:00:00.000+02:00');

console.log('April days', start.plus({ months: 1 }).startOf('month').diff(start, 'days').days);
console.log('April months', start.plus({ months: 1 }).startOf('month').diff(start, 'months').months);

// 25 days in the first month or around 0.8333333333333334 months
// (25 / 30) ~= 0,833333333

console.log('May days', end.diff(start.plus({ months: 1 }).startOf('month'), 'days').days);
console.log('May months', end.diff(start.plus({ months: 1 }).startOf('month'), 'months').months);

// 31 days in may, that is exactly 1 month

console.log('Total days', end.diff(start, 'days').days);
console.log('Total months', end.diff(start, 'months').months);

// 56 days in the full period or around 1.838709677419355 months
```

The sum of the previous two diffs is around 1.833333333 months.
The difference is quite significant with 0.005376344 months!

Same behaviour for:
  * the `.length()` method on an Interval generated by `Interval.fromDateTimes`
  * UTC ISO strings.

Installed packages:
* full-icu@1.3.0
* luxon@1.16.1

**Comment by icambron** [comment]
This is not a bug; it's a variant of [losing information](https://moment.github.io/luxon/docs/manual/math.html#losing-information). Months aren't fixed amount of time, so by splitting up the month you messed with Luxon's ability to do calendar math:

 * `end.diff(start, 'months')` says "April 5 to May 5 is one month (irrespective of the number of days). A second month would be June 5, and we only make it 26/31 of the way there, so we have 1 + 26/31 months = 1.839 months.
 * April 5 - May 1 is 25/30 of the way to May 5, and May 1 - June 1 is one month. So 1 + 25/30 = 1.833 months

In short, you can't split things up like that. If you want to use a consistent length of months, just do the diff in milliseconds and convert it to months (i.e. lose the information on purpose):

```js
end.diff(start).as('months') //=> 1.8666666666666667, based on 1 month = 30 days
end.diff(start, 'milliseconds', { conversionAccuracy: 'longterm' }).as('months') // => 1.839873508696277, based on 1 month = 30.436875 days (average for months overall)
```

**Comment by captainmuppet** [comment]
Ok, thank you!

My goal is to split up a price evenly between the months of a year, and then do the same for the number of days it was active.

$PRICE / 365 has issues because of leap years.

The correct solution would be to use ($PRICE / 12) * ($END - $START in months), but in order for this to work, the length function would have to calculate the lengths on a month-by-month basis. The advantage over fixed-length months is that you don't need a full (two months / year) to compensate for errors in the previous calculations.

Is there a way of doing this in luxon without explicitly splitting the whole interval into months?

**Comment by icambron** [comment]
IMO, `totalAmount/end.diff(start, 'months')` is exactly what you want for that. As it applied to dividing up payments, it divides up the amount across all the months, but prorating the last month by the faction of the month that falls within the usage period. Luxon did all the work of thinking through the various lengths of the months. April 5 - May 5 is the first month and May 5 - June 1 is the partial last month.

What your example does differently is to do the prorating on the beginning (filling out the first month) instead of on the end. Ideally, diffing backwards (`start.diff(end, "months")`) would do what you want (June 1 backward to May 1 as the first day, then May 1 to April 5 as the fraction), but that isn't currently how Luxon works (instead, it flips the dates, does the calculation, and takes its negative).

Finally, what you did (meaning, compute the diff to the first month boundary, and then diff the rest) is not actually wrong if that's the semantics you want. So if prorating the balance of the first month is what you want to do, I think you have it right. You just shouldn't expect it to equal the overall diff, because Luxon makes the opposite choice about how to do the fractional part.
